### PR TITLE
Update how to extend Ballerina for specific webhooks

### PR DESCRIPTION
**Pull request by MaryamZi**
Opening a new PR in place of #1338 

## Purpose
This PR updates "How to Extend Ballerina" with specific webhook implementation details.

## Depends on PRs
https://github.com/ballerina-platform/ballerina-lang/pull/12938
https://github.com/wso2-ballerina/module-github/pull/66

### CODE REVIEW

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 1]
```diff
@@ -952,3 +954,181 @@ Kubernetes
...
+
+1. Create a Ballerina module in a Ballerina project.
+
+2. Create a new `listener` wrapping the generic `websub:Listener`.
+
+3. Implement the `__init()` function specifying the mapping between possible notifications and resources as `extensionConfig`.
+
+4. Implement the `__attach()`, `__start()` and `__stop()` functions calling the same functions on the generic `websub:Listener`.
+
+5. Build the module and push it into a registry for usage by others.
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
5. Build the module and push it into a registry to be used by others.
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 2]
```diff
@@ -952,3 +954,181 @@ Kubernetes
...
+    public function __stop() returns error?;
+};
+```
+
+### Implement the `__init()` function
+
+For the GitHub webhook, the mapping between events and resources could be based either only on a header or on both a header and the payload.
+
+e.g.,
+- when the repository is starred, the content delivery request received will have the header `X-GitHub-Event` with the value set to "watch".
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
- When the repository is starred, the content delivery request received will have the header `X-GitHub-Event` with the value set to "watch".
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 3]
```diff
@@ -952,3 +954,181 @@ Kubernetes
...
+- when an issue is opened, the content delivery request received will have the header `X-GitHub-Event` with the
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
- When an issue is opened, the content delivery request received will have the header `X-GitHub-Event` with the 
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 4]
```diff
@@ -952,3 +954,181 @@ Kubernetes
...
+import ballerina/io;
+import ballerina/websub;
+import wso2/github;
+
+// Initiate a new listener of type `github:WebhookListener`.
+listener github:WebhookListener githubListener = new(8080);
+
+// Attach a service accepting ping, issue opened and repository starred notifications.
+@websub:SubscriberServiceConfig {
+   path:"/webhook",
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
   path: "/webhook",
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 5]
```diff
@@ -952,3 +954,181 @@ Kubernetes
...
+   subscribeOnStartUp: true,
+   hub: github:HUB,
+   topic: "https://github.com/<GH_USERNAME>/<GH_REPO_NAME>/events/*.json", // subscribe to notifications for all events
+   secret: "<SECRET>",
+   callback: "<CALLBACK_URL>", // only needs to be specified if not http(s)://<HOST>:<PORT>/<path>
+   subscriptionClientConfig: {
+        auth: {
+            scheme:http:OAUTH2,
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
            scheme: http:OAUTH2,
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 6]
```diff
@@ -952,3 +954,181 @@ Kubernetes
...
+            accessToken:"<GH_ACCESS_TOKEN>"
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
            accessToken: "<GH_ACCESS_TOKEN>"
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 7]
```diff
@@ -952,3 +954,181 @@ Kubernetes
...
+- when the repository is starred, the content delivery request received will have the header `X-GitHub-Event` with the value set to "watch".
+- when an issue is opened, the content delivery request received will have the header `X-GitHub-Event` with the 
+value set to "issues", and the `json` payload will contain the value "opened" for the key "action".
+
+The implementation of the `__init()` function initializes the `websub:Listener` specifying the mapping to the resources as the `extensionConfig`.
+```ballerina
+const string TOPIC_HEADER = "X-GitHub-Event";
+
+final map<(string, typedesc)> GITHUB_TOPIC_HEADER_RESOURCE_MAP = {
+    "ping" : ("onPing", PingEvent),
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
    "ping": ("onPing", PingEvent),
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 8]
```diff
@@ -952,3 +954,181 @@ Kubernetes
...
+    ...
+    "watch" : ("onWatch", WatchEvent)
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
    "watch": ("onWatch", WatchEvent)
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 9]
```diff
@@ -952,3 +954,181 @@ Kubernetes
...
+};
+
+final map<map<map<(string, typedesc)>>> GITHUB_TOPIC_HEADER_AND_PAYLOAD_KEY_RESOURCE_MAP = {
+    "create" : {
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
    "create": {
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 10]
```diff
@@ -952,3 +954,181 @@ Kubernetes
...
+        "ref_type" : {
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
        "ref_type": {
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 11]
```diff
@@ -952,3 +954,181 @@ Kubernetes
...
+            "repository" : ("onCreateRepository", CreateEvent)
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
            "repository": ("onCreateRepository", CreateEvent)
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 12]
```diff
@@ -952,3 +954,181 @@ Kubernetes
...
+            ...
+        }
+    },
+    "issues" : {
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
    "issues": {
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 13]
```diff
@@ -952,3 +954,181 @@ Kubernetes
...
+        "action" : {
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
        "action": {
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 14]
```diff
@@ -952,3 +954,181 @@ Kubernetes
...
+            "opened" : ("onIssuesOpened", IssuesEvent),
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
            "opened": ("onIssuesOpened", IssuesEvent),
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 15]
```diff
@@ -952,3 +954,181 @@ Kubernetes
...
+            ...
+            "closed" : ("onIssuesClosed", IssuesEvent),
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
            "closed": ("onIssuesClosed", IssuesEvent),
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md (hunk 1 above)
Comment (MaryamZi) [reply website/two-column-pages/docs/learn/how-to-extend-ballerina.md]: These are the same steps as for client connectors.

IMO, "to be used" could be misleading, not clear if we are talking about the registry itself.

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 16]
```diff
@@ -1,11 +1,13 @@
 # How to Extend Ballerina
 
-Developers and third parties can extend the behavior of Ballerina and collate the customizations for use by others. There are two ways to customize the behavior of Ballerina:
+Developers and third parties can extend the behavior of Ballerina and collate the customizations for use by others. There are three ways to customize the behavior of Ballerina:
 
 1. Collate and distribute new client connectors to third party endpoints, such as databases, infrastructure and APIs.
 
 2. Add new annotations to Ballerina source files so that the compiler can act to alter binaries and generate artifacts.
 
+3. Collate and distribute new webhook callback services, that programmatically subscribe to accept content delivery requests on the occurrence of particular events.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
3. Collate and distribute new webhook callback services, that programmatically subscribe to accept content delivery requests on the occurrence of particular events.
```

You don't use a comma after "that".

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 17]
```diff
@@ -915,3 +917,182 @@ Kubernetes
 1. The [Ballerina file defining the annotation](https://github.com/ballerinax/kubernetes/blob/master/kubernetes-extension/src/main/ballerina/ballerinax/kubernetes/annotation.bal).
 
 2. The [Java code with the builder extension](https://github.com/ballerinax/kubernetes/tree/master/kubernetes-extension/src/main/java/org/ballerinax/kubernetes).
+
+## Create Webhook Callback Services
+
+A webhook callback service type is introduced by implementing a new `listener` type wrapping the generic 
+`websub:Listener`. The implementation should define a mapping between an indicator (in the content delivery requests) 
+and the resources (to which the requests need to be dispatched based on the value of the indicator).
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]: Why do we have a gap between this and the previous line?

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 18]
```diff
@@ -915,3 +917,182 @@ Kubernetes
...
+ 
+This indicator, based on which the extension is done, could be:
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
This indicator could be one of the following  based on the extension done:
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 19]
```diff
@@ -915,3 +917,182 @@ Kubernetes
 1. The [Ballerina file defining the annotation](https://github.com/ballerinax/kubernetes/blob/master/kubernetes-extension/src/main/ballerina/ballerinax/kubernetes/annotation.bal).
 
 2. The [Java code with the builder extension](https://github.com/ballerinax/kubernetes/tree/master/kubernetes-extension/src/main/java/org/ballerinax/kubernetes).
+
+## Create Webhook Callback Services
+
+A webhook callback service type is introduced by implementing a new `listener` type wrapping the generic 
+`websub:Listener`. The implementation should define a mapping between an indicator (in the content delivery requests)
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
`websub:Listener`. The implementation should define a mapping between an indicator (in the content delivery requests).
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 20]
```diff
@@ -915,3 +917,182 @@ Kubernetes
...
+This indicator, based on which the extension is done, could be:
+- A request header 
+- The payload - the value for a particular key in the JSON payload
+- A request header and the payload (combination of the above)
+ 
+The [ballerina/websub Module.md](stdlib/websub/src/main/ballerina/websub/Module.md) explains the extension points in detail.
+ 
+You can create and share your own webhook callback service types as Ballerina modules, which you push into a Ballerina registry (such as Ballerina Central).
+
+To create a webhook callback service type, you:
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
Follow the steps below to create a webhook callback service type.
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md (hunk 1 above)
Comment (praneesha) [reply website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
5. Build the module and push it into a registry for others to use it.
```
Bette to say the idea directly and simply.

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 21]
```diff
@@ -915,3 +917,182 @@ Kubernetes
...
+
+4. Implement the `__attach()`, `__start()` and `__stop()` functions calling the same functions on the generic `websub:Listener`.
+
+5. Build the module and push it into a registry for usage by others.
+
+### The GitHub Webhook
+
+A GitHub webhook implementation is made available by WSO2 as the module `wso2/githubwebhook3`.
+
+This webhook can be used by anyone, by importing the `wso2/githubwebhook3` module.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
This webhook can be used by anyone by importing the `wso2/githubwebhook3` module.
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 22]
```diff
@@ -915,3 +917,182 @@ Kubernetes
...
+```ballerina
+import ballerina/http;
+import ballerina/io;
+import ballerina/websub;
+import wso2/githubwebhook3;
+
+// Initiate a new listener of type `githubwebhook3:WebhookListener`.
+listener githubwebhook3:WebhookListener githubListener = new(8080);
+
+// Attach a service accepting ping, issue opened and repository starred notifications.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
// Attach a service accepting 'ping', 'issue opened' and 'repository-starred notifications'.
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 23]
```diff
@@ -915,3 +917,182 @@ Kubernetes
...
+    public function __stop() returns error?;
+};
+```
+
+### Implement the `__init()` function
+
+For the GitHub webhook, the mapping between events and resources could be based either only on a header or on both a header and the payload.
+
+e.g.,
+- When the repository is starred, the content delivery request received will have the header `X-GitHub-Event` with
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
- When the repository is starred, the content-delivery request received will have the header `X-GitHub-Event` with 
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 24]
```diff
@@ -915,3 +917,182 @@ Kubernetes
...
+```
+
+### Implement the `__init()` function
+
+For the GitHub webhook, the mapping between events and resources could be based either only on a header or on both a header and the payload.
+
+e.g.,
+- When the repository is starred, the content delivery request received will have the header `X-GitHub-Event` with 
+the value set to "watch".
+- When an issue is opened, the content delivery request received will have the header `X-GitHub-Event` with the
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
- When an issue is opened, the content-delivery request received will have the header `X-GitHub-Event` with the 
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 25]
```diff
@@ -915,3 +917,182 @@ Kubernetes
...
+        }
+        sseConfig.httpServiceSecureSocket = config["httpServiceSecureSocket"];
+    }
+    self.websubListener = new(port, config = sseConfig);
+}
+```
+
+### Implement the `__attach()`, `__start()` and `__stop()` functions
+
+The implementions of these functions only need to call the same functions on the wrapped `websub:Listener`.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
The implementations of these functions need to call only the functions of the wrapped `websub:Listener`.
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 26]
```diff
@@ -915,3 +917,182 @@ Kubernetes
...
+
+3. Implement the `__init()` function specifying the mapping between possible notifications and resources as `extensionConfig`.
+
+4. Implement the `__attach()`, `__start()` and `__stop()` functions calling the same functions on the generic `websub:Listener`.
+
+5. Build the module and push it into a registry for usage by others.
+
+### The GitHub Webhook
+
+A GitHub webhook implementation is made available by WSO2 as the module `wso2/githubwebhook3`.
```
[the remark this answers is not in the document]
Comment (MaryamZi) [reply website/two-column-pages/docs/learn/how-to-extend-ballerina.md]: The module itself contains the implementation. How about
```suggestion
A GitHub webhook implementation is made available by WSO2 as the `wso2/githubwebhook3` module.
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md (hunk 25 above)
Comment (MaryamZi) [reply website/two-column-pages/docs/learn/how-to-extend-ballerina.md]: Fixed the typo. The functions are called on the wrapped object (`listener.__init()`).

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md (hunk 18 above)
Comment (praneesha) [reply website/two-column-pages/docs/learn/how-to-extend-ballerina.md]: +1

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md (hunk 25 above)
Comment (praneesha) [reply website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
The implementations of these functions only need to call the corresponding functions of the wrapped `websub:Listener`.
```

**Pull request merged**